### ISSUE TITLE
An error inside the WebSocket handler results in "Reply was already sent" and not the actual error

### ISSUE THREAD

**Issue by Prinzhorn** [fastify/fastify-websocket#119]
## 🐛 Bug Report

https://github.com/fastify/fastify-websocket/issues/103#issuecomment-797639332

Errors inside the handler are currently swallowed, which for me is unexpected behavior and inconsistent with the behavior for regular non-ws routes.

## To Reproduce

```js
'use strict'

const fastify = require('fastify')({
  logger: true
})

fastify.register(require('fastify-websocket'))

fastify.get('/ws', { websocket: true }, () => {
  throw new Error('demonstration')
})

// Declare a route
fastify.get('/', (request, reply) => {
  throw new Error('demonstration')
})

fastify.listen(3000, err => {
  if (err) {
    fastify.log.error(err)
    process.exit(1)
  }
})
```

`wscat --connect ws://localhost:3000/ws` results in :

```
{
  "level": 40,
  "time": 1615569367599,
  "pid": 28959,
  "hostname": "alex-desktop",
  "reqId": "req-1",
  "err": {
    "type": "FastifyError",
    "message": "Reply was already sent.",
    "stack": "FastifyError: Reply was already sent.\n    at _Reply.Reply.send (/home/alex/src/issues/issue-256-websocket-swallows-error/node_modules/fastify/lib/reply.js:118:26)\n    at preHandlerCallback (/home/alex/src/issues/issue-256-websocket-swallows-error/node_modules/fastify/lib/handleRequest.js:126:11)\n    at preValidationCallback (/home/alex/src/issues/issue-256-websocket-swallows-error/node_modules/fastify/lib/handleRequest.js:107:5)\n    at handler (/home/alex/src/issues/issue-256-websocket-swallows-error/node_modules/fastify/lib/handleRequest.js:70:7)\n    at handleRequest (/home/alex/src/issues/issue-256-websocket-swallows-error/node_modules/fastify/lib/handleRequest.js:18:5)\n    at runPreParsing (/home/alex/src/issues/issue-256-websocket-swallows-error/node_modules/fastify/lib/route.js:421:5)\n    at Object.routeHandler [as handler] (/home/alex/src/issues/issue-256-websocket-swallows-error/node_modules/fastify/lib/route.js:379:7)\n    at Router.lookup (/home/alex/src/issues/issue-256-websocket-swallows-error/node_modules/find-my-way/index.js:339:14)\n    at Server.<anonymous> (/home/alex/src/issues/issue-256-websocket-swallows-error/node_modules/fastify-websocket/index.js:50:15)\n    at Server.emit (events.js:315:20)",
    "name": "FastifyError",
    "code": "FST_ERR_REP_ALREADY_SENT",
    "statusCode": 500
  },
  "msg": "Reply already sent"
}
```

`curl  http://localhost:3000` :

```
{
  "level": 50,
  "time": 1615569798536,
  "pid": 30071,
  "hostname": "alex-desktop",
  "reqId": "req-1",
  "req": {
    "method": "GET",
    "url": "/",
    "hostname": "localhost:3000",
    "remoteAddress": "127.0.0.1",
    "remotePort": 50970
  },
  "res": { "statusCode": 500 },
  "err": {
    "type": "Error",
    "message": "demonstration",
    "stack": "Error: demonstration\n    at Object.<anonymous> (/home/alex/src/issues/issue-256-websocket-swallows-error/index.js:16:9)\n    at Object.routeOptions.handler (/home/alex/src/issues/issue-256-websocket-swallows-error/node_modules/fastify-websocket/index.js:125:24)\n    at preHandlerCallback (/home/alex/src/issues/issue-256-websocket-swallows-error/node_modules/fastify/lib/handleRequest.js:124:28)\n    at preValidationCallback (/home/alex/src/issues/issue-256-websocket-swallows-error/node_modules/fastify/lib/handleRequest.js:107:5)\n    at handler (/home/alex/src/issues/issue-256-websocket-swallows-error/node_modules/fastify/lib/handleRequest.js:70:7)\n    at handleRequest (/home/alex/src/issues/issue-256-websocket-swallows-error/node_modules/fastify/lib/handleRequest.js:18:5)\n    at runPreParsing (/home/alex/src/issues/issue-256-websocket-swallows-error/node_modules/fastify/lib/route.js:421:5)\n    at Object.routeHandler [as handler] (/home/alex/src/issues/issue-256-websocket-swallows-error/node_modules/fastify/lib/route.js:379:7)\n    at Router.lookup (/home/alex/src/issues/issue-256-websocket-swallows-error/node_modules/find-my-way/index.js:339:14)\n    at Server.emit (events.js:315:20)"
  },
  "msg": "demonstration"
}
```

For the ws handler the information about the original `new Error('demonstration')` is completely lost.

## Expected behavior

Errors in my code are logged and I get a chance to track them down.

## Your Environment

- *node version*: v14.16.0
- *fastify version*: 3.14.0, ws: 3.1.0
- *os*: Ubuntu

**Comment by mcollina** [comment]
Thanks for reporting!  We are missing a try-catch around https://github.com/fastify/fastify-websocket/blob/9b0a134d6593d6178152983068d31065bee0ee02/index.js#L114-L118.

Everything would work correctly if you make your route `async`.

**Comment by mcollina** [comment]
Would you like to send a Pull Request to address this issue? Remember to add unit tests.

**Comment by wpbjiuy** [comment]
I also encountered the same problem. Have you solved it?

**Comment by Prinzhorn** [comment]
@wpbjiuy fixed in v3.2.0

**Comment by wpbjiuy** [comment]
> @wpbjiuy fixed in v3.2.0
I installed v3.2.0, but it is still reporting an error
`
err: {
      "type": "FastifyError",
      "message": "Reply was already sent.",
      "stack":
          FastifyError: Reply was already sent.
              at _Reply.Reply.send (E:\shiKe\serve\node_modules\fastify\lib\reply.js:118:26)
              at preHandlerCallback (E:\shiKe\serve\node_modules\fastify\lib\handleRequest.js:126:11)
              at preValidationCallback (E:\shiKe\serve\node_modules\fastify\lib\handleRequest.js:107:5)
              at handler (E:\shiKe\serve\node_modules\fastify\lib\handleRequest.js:70:7)
              at handleRequest (E:\shiKe\serve\node_modules\fastify\lib\handleRequest.js:18:5)
              at runPreParsing (E:\shiKe\serve\node_modules\fastify\lib\route.js:421:5)
              at next (E:\shiKe\serve\node_modules\fastify\lib\hooks.js:158:7)
              at handleResolve (E:\shiKe\serve\node_modules\fastify\lib\hooks.js:175:5)
              at processTicksAndRejections (internal/process/task_queues.js:97:5)
      "name": "FastifyError",
      "code": "FST_ERR_REP_ALREADY_SENT",
      "statusCode": 500
    }
    reqId: "req-3"
`

**Comment by Prinzhorn** [comment]
@wpbjiuy I cannot reproduce this anymore with the code I posted above. Can you provide a similar self-contained example that shows the problem?
